Log IP on SSH authentication failure for Built-in SSH server

elesiuta opened the pull request:
targets https://github.com/go-gitea/gitea/issues/13094

Line 127 logs the IP address when connecting with a key from a user that does not have permission for the repo they're trying to clone or the repo does not exist.
Line 188 logs the IP address when connecting with a key not recognized by the server.
I'm not sure which cases Line 131 will be hit (if any and if Line 127 already covers this), but including it seemed prudent.

The regex from this [guide](https://docs.gitea.io/en-us/fail2ban-setup/) could also be updated to omit the 'for' so it matches this along with failed logins, should I include that in this pull request or make a new one? It could also be updated to match "invalid credentials from" in the log when cloning over HTTP.

lafriks commented:
Yes, please update docs in this PR

zeripath commented on modules/ssh/ssh.go:
```diff
@@ -124,11 +124,11 @@ func sessionHandler(session ssh.Session) {
 	err = cmd.Wait()
 	if err != nil {
-		log.Error("SSH: Wait: %v", err)
+		log.Error("SSH: Wait: %v Failed authentication attempt from %s", err, session.RemoteAddr())
```
Are you sure that every error from cmd wait will be due to an authentication failure?

elesiuta replied:
Nope, maybe a check if err is `exit status 1` but then err could also be other things and still be an authentication failure and I don't want to miss those. I'm not too familiar with the code base yet and only started using gitea this week. 

Another option is prepend it with 'Possible' or omit 'Failed authentication attempt' entirely, leaving just the IP and leave it to the user to decide whether to include this in fail2ban.

zeripath commented on modules/ssh/ssh.go:
```diff
@@ -124,11 +124,11 @@ func sessionHandler(session ssh.Session) {
 
 	if err := session.Exit(getExitStatusFromError(err)); err != nil {
-		log.Error("Session failed to exit. %s", err)
+		log.Error("Session failed to exit. %s Failed authentication attempt from %s", err, session.RemoteAddr())
```
Similarly?

elesiuta replied:
I'm even less sure about this one, wasn't able to find an input that triggers this in my testing.

zeripath commented:
Ok so I think this is all in the wrong place.

By the time the session handler is involved they have authenticated with the server as using a public key. 

You presumably want to catch the people who are failing to authenticate.

In terms of errors from the session handler - those are more authorization errors and you'd be better off logging them in routers/private/serv.go

elesiuta commented:
> Ok so I think this is all in the wrong place.
> 
> By the time the session handler is involved they have authenticated with the server as using a public key.
> 
> You presumably want to catch the people who are failing to authenticate.
> 
> In terms of errors from the session handler - those are more authorization errors and you'd be better off logging them in routers/private/serv.go

Thank you! You're right, that place is much better. I've updated it to reflect this, and also added one more location to log the remote IP address, when trying to clone over HTTP for a user that does not exist.

6543 commented on docs/content/doc/usage/fail2ban-setup.en-us.md:
```diff
@@ -18,20 +18,26 @@ menu:
-Gitea returns an HTTP 200 for bad logins in the web logs, but if you have logging options on in
-`app.ini`, then you should be able to go off of `log/gitea.log`, which gives you something like this
-on a bad authentication:
+Gitea returns an HTTP 200 for bad logins in the web logs, but if you have logging options on in 
+`app.ini`, then you should be able to go off of `log/gitea.log`, which gives you something like this 
```
```suggestion
Gitea returns an HTTP 200 for bad logins in the web logs, but if you have logging options on in
`app.ini`, then you should be able to go off of `log/gitea.log`, which gives you something like this
```

6543 reviewed (changes requested):
as per @zeripath `...By the time the session handler is involved they have authenticated with the server as using a public key...`

6543 commented:
:rocket:
